File: version_2_Julia/tools/test_cost_at_theta0.jl
#!/usr/bin/env julia
# Bit-for-bit cost comparison Python vs Julia — Julia side.
#
# Standing-hypothesis test from docs/julia_fsa_writeup.pdf §2.7 item 1:
# fix θ (start with θ = 0), use a saved CRN noise grid, print the
# per-trial A_acc that the Julia GPU kernel computes. Compares to the
# Python twin's saved per-trial values.
#
# The on-disk kernel in models/fsa_high_res/gpu_control.jl is
# Eq.37-only (no cost_state_weights), so:
#   • lam_F = 0  → per-trial cost = -A_acc      → A_acc = -cost
#   • lam_F = 1  → per-trial cost = -A_acc + barrier_acc
# At θ=0 (Φ ≡ 1.0) Python finds barrier_acc ≡ 0 (F never exceeds F_max),
# so a single lam_F=0 run is enough to extract A_acc.
#
# Prereq: run test_cost_at_theta0.py first to create
#   /tmp/crn_noise_seed42.npy
#   /tmp/python_cost_at_theta0.npz
#
# Run:
#   cd version_2_Julia
#   julia --project=. tools/test_cost_at_theta0.jl

ENV["FSA_STEP_MINUTES"] = "15"

using CUDA, Random, Statistics, Printf, NPZ

const REPO_ROOT = abspath(joinpath(@__DIR__, ".."))
include(joinpath(REPO_ROOT, "models", "fsa_high_res", "FSAHighRes.jl"))
using .FSAHighRes
using .FSAHighRes.GPUControl: FSAControlGPUTarget, gpu_cost_log_density_batched
using .FSAHighRes.Dynamics: TRUTH_PARAMS, A_TYP, F_TYP


# ── Pure-Julia fp64 CPU mirror of fsa_cost_kernel! ──────────────────────
# Used to formally test §2.7 item 2 (fp32 A-accumulation cancellation).
# Algorithm is byte-identical to the GPU kernel; only the dtype differs.
function cpu_fp64_per_trial_cost(theta::Vector{Float64}, lam_F::Float64,
                                  noise_fp64::Array{Float64,3};
                                  init_state = [0.05, 0.30, 0.10],
                                  Phi_max = 3.0, Phi_default = 1.0,
                                  F_max = 0.40, n_substeps = 4,
                                  dt = 1.0/96.0, T_days = 14)
    n_inner   = size(noise_fp64, 1)
    n_steps   = size(noise_fp64, 2)
    n_anchors = length(theta)
    @assert size(noise_fp64) == (n_inner, n_steps, 3)

    p = TRUTH_PARAMS
    p_ratio = Phi_default / Phi_max
    c_Phi   = log(p_ratio / (1.0 - p_ratio))

    # RBF design matrix (same as gpu_control.jl).
    T_total = n_steps * dt
    t_grid  = collect(0:n_steps-1) .* dt
    anchors = collect(range(0.0, T_total; length=n_anchors))
    σ_rbf   = T_total / n_anchors
    rbf = [exp(-0.5 * ((t_grid[k] - anchors[j]) / σ_rbf)^2)
           for k in 1:n_steps, j in 1:n_anchors]

    sub_dt  = dt / n_substeps
    sqrt_dt = sqrt(dt)
    eps_B   = 1e-4

    out = zeros(Float64, n_inner)
    for t in 1:n_inner
        B = init_state[1]; F = init_state[2]; A = init_state[3]
        A_acc = 0.0; barrier_acc = 0.0

        for k in 1:n_steps
            raw = c_Phi
            for a in 1:n_anchors
                raw += theta[a] * rbf[k, a]
            end
            Phi_t = Phi_max / (1.0 + exp(-raw))

            # PRE-step accumulation (matches kernel).
            A_acc += A * dt
            barrier_acc += max(F - F_max, 0.0)^2 * dt

            # Substepped EM — simultaneous update (top-of-loop reads only).
            for _ in 1:n_substeps
                F_dev = F - F_TYP
                mu_bif = p.mu_0 + p.mu_B * B - p.mu_F * F - p.mu_FF * F_dev * F_dev
                a_factor_B = (1.0 + p.epsilon_A * A) / (1.0 + p.epsilon_A * A_TYP)
                a_factor_F = (1.0 + p.lambda_A  * A) / (1.0 + p.lambda_A  * A_TYP)
                dB = p.kappa_B * a_factor_B * Phi_t - B / p.tau_B
                dF = p.kappa_F * Phi_t - a_factor_F / p.tau_F * F
                dA = mu_bif * A - p.eta * A * A * A
                B += sub_dt * dB
                F += sub_dt * dF
                A += sub_dt * dA
            end

            B_cl = max(eps_B, min(1.0 - eps_B, B))
            F_cl = max(0.0, F)
            A_cl = max(0.0, A)
            σB_eff = p.sigma_B * sqrt(B_cl * (1.0 - B_cl))
            σF_eff = p.sigma_F * sqrt(F_cl)
            σA_eff = p.sigma_A * sqrt(A_cl)

            B += σB_eff * sqrt_dt * noise_fp64[t, k, 1]
            F += σF_eff * sqrt_dt * noise_fp64[t, k, 2]
            A += σA_eff * sqrt_dt * noise_fp64[t, k, 3]

            B = B < 0.0 ? -B : (B > 1.0 ? 2.0 - B : B)
            F = abs(F)
            A = abs(A)
        end

        out[t] = -A_acc + lam_F * barrier_acc
    end
    return out
end

# ── Config — must match Python twin exactly ─────────────────────────────
const N_INNER      = 32
const T_DAYS       = 14
const BINS_PER_DAY = 96
const N_STEPS      = T_DAYS * BINS_PER_DAY     # 1344
const DT           = 1.0 / BINS_PER_DAY
const N_SUBSTEPS   = 4
const N_ANCHORS    = 8
const F_MAX        = 0.40
const PHI_MAX      = 3.0
const PHI_DEFAULT  = 1.0
const SEED         = 42
const M            = 1                          # one chain per call
const M_MAX        = 1


# ── Load Python's CRN noise (fp64) and cast to fp32 ─────────────────────
const NOISE_PATH = "/tmp/crn_noise_seed$(SEED).npy"
const PY_PATH    = "/tmp/python_cost_at_theta0.npz"

isfile(NOISE_PATH) || error("Run test_cost_at_theta0.py first to create $NOISE_PATH")
isfile(PY_PATH)    || error("Run test_cost_at_theta0.py first to create $PY_PATH")

noise_fp64 = NPZ.npzread(NOISE_PATH)
@assert size(noise_fp64) == (N_INNER, N_STEPS, 3)
noise_fp32 = Float32.(noise_fp64)
println("Loaded noise → $NOISE_PATH  shape=$(size(noise_fp32))  dtype=$(eltype(noise_fp32))")


"""
    eval_per_trial_cost(theta::Vector{Float64}, lam_F::Float64) -> Vector{Float32}

Build a fresh FSAControlGPUTarget at the given lam_F, override its
RNG-generated noise grid with the Python-saved CRN, evaluate at the
given θ vector, and return the per-trial cost (length N_INNER) for the
single chain.
"""
function eval_per_trial_cost(theta::Vector{Float64}, lam_F::Float64)
    @assert length(theta) == N_ANCHORS
    target = FSAControlGPUTarget(
        n_inner   = N_INNER, M_max = M_MAX,
        n_steps   = N_STEPS, n_anchors = N_ANCHORS, n_substeps = N_SUBSTEPS,
        dt        = DT,
        F_max     = F_MAX, Phi_max = PHI_MAX, Phi_default = PHI_DEFAULT,
        lam_F     = lam_F,
        sigma_prior = 1.5,
        params    = TRUTH_PARAMS,
        init_state = [0.05, 0.30, 0.10],
        noise_seed = SEED,                # ignored — overwritten below
    )
    # Override the RNG-generated noise with the saved Python CRN (fp64 → fp32).
    copyto!(target.fixed_w, noise_fp32)

    theta_unc = reshape(theta, 1, N_ANCHORS)
    _ = gpu_cost_log_density_batched(target, theta_unc)
    cost_cpu = Array(view(target.cost_per_thread, 1:M*N_INNER))
    return reshape(cost_cpu, N_INNER, M)[:, 1]
end


# ── Helper to print the θ-vs-θ comparison block ─────────────────────────
function compare_at_theta(label::String, theta::Vector{Float64},
                          A_accs_py::Vector{Float64},
                          barrier_accs_py::Vector{Float64})
    println("\nRunning Julia GPU kernel at $label with saved Python CRN…")
    cost_lamF0 = eval_per_trial_cost(theta, 0.0)    # = -A_acc
    cost_lamF1 = eval_per_trial_cost(theta, 1.0)    # = -A_acc + barrier_acc

    A_accs_jl       = Float64.(-cost_lamF0)
    barrier_accs_jl = Float64.(cost_lamF1 - cost_lamF0)

    println()
    println("="^78)
    println("$label per-trial cost components — Julia (fp32) vs Python (fp64), same CRN")
    println("="^78)
    @printf("  θ = %s\n", string(theta))
    @printf("%-13s  %-22s  %-22s  %-9s\n",
            "Quantity", "Julia mean ± std", "Python mean ± std", "rel.diff")
    println("-"^78)
    for (name, jl, py) in [
        ("A_acc",       A_accs_jl,       A_accs_py),
        ("barrier_acc", barrier_accs_jl, barrier_accs_py),
    ]
        jl_m = mean(jl); py_m = mean(py)
        jl_s = std(jl);  py_s = std(py)
        rel  = abs(jl_m - py_m) / max(abs(py_m), 1e-12)
        @printf("%-13s  %.5f ± %.5f      %.5f ± %.5f      %6.3f%%\n",
                name, jl_m, jl_s, py_m, py_s, 100*rel)
    end

    println()
    println("First 5 per-trial A_acc:")
    @printf("  Julia:  %s\n", join([@sprintf("%.5f", v) for v in A_accs_jl[1:5]], ", "))
    @printf("  Python: %s\n", join([@sprintf("%.5f", v) for v in A_accs_py[1:5]], ", "))
    @printf("  diff:   %s\n",
            join([@sprintf("%+.5f", A_accs_jl[i] - A_accs_py[i]) for i in 1:5], ", "))

    diffs    = abs.(A_accs_jl .- A_accs_py)
    rel_diffs = diffs ./ max.(abs.(A_accs_py), 1e-12)
    @printf("\nPer-trial |Jl-Py| / |Py|  A_acc:        max = %.4f%%   mean = %.4f%%\n",
            100*maximum(rel_diffs), 100*mean(rel_diffs))

    if any(barrier_accs_py .> 1e-12) || any(barrier_accs_jl .> 1e-12)
        bdiffs    = abs.(barrier_accs_jl .- barrier_accs_py)
        brel      = bdiffs ./ max.(abs.(barrier_accs_py), 1e-12)
        @printf("Per-trial |Jl-Py| / |Py|  barrier_acc:  max = %.4f%%   mean = %.4f%%\n",
                100*maximum(brel), 100*mean(brel))
    end

    @printf("Mean ∫A/T  Julia: %.4f   Python: %.4f\n",
            mean(A_accs_jl) / (N_STEPS * DT), mean(A_accs_py) / (N_STEPS * DT))

    return A_accs_jl, barrier_accs_jl
end


# ── Run both θ values ───────────────────────────────────────────────────
py = NPZ.npzread(PY_PATH)

# θ=0 (Banister baseline)
theta0 = zeros(Float64, N_ANCHORS)
compare_at_theta("θ=0", theta0,
                  Vector{Float64}(py["A_acc"]),
                  Vector{Float64}(py["barrier_acc"]))

# θ_RO (recovery→overload anti-symmetric, writeup §2.4)
theta_RO = [-1.5, -1.5, -1.5, 0.0, 0.0, +1.5, +1.5, +1.5]
compare_at_theta("θ_RO", theta_RO,
                  Vector{Float64}(py["A_acc_RO"]),
                  Vector{Float64}(py["barrier_acc_RO"]))


# ── §2.7 item 2: fp64 A-accumulation cross-check (CPU mirror) ───────────
# If fp32 cancellation in ∫A over 5376 substeps were material, the GPU-fp32
# A_acc would diverge from the CPU-fp64 mirror at the same θ + same noise.
# We expect three-way agreement (GPU-fp32 ≈ CPU-fp64 ≈ Python-fp64).
println("\n" * "="^78)
println("§2.7 item 2:  fp64 CPU mirror — rules out fp32 cancellation in ∫A")
println("="^78)
for (label, theta, py_key) in [
    ("θ=0",   theta0,   "A_acc"),
    ("θ_RO",  theta_RO, "A_acc_RO"),
]
    A_jl_gpu_fp32 = -Float64.(eval_per_trial_cost(theta, 0.0))
    A_jl_cpu_fp64 = -cpu_fp64_per_trial_cost(theta, 0.0, noise_fp64)
    A_py_fp64     = Vector{Float64}(py[py_key])

    rel_gpu_vs_cpu = abs.(A_jl_gpu_fp32 .- A_jl_cpu_fp64) ./ max.(abs.(A_jl_cpu_fp64), 1e-12)
    rel_cpu_vs_py  = abs.(A_jl_cpu_fp64 .- A_py_fp64)     ./ max.(abs.(A_py_fp64),     1e-12)

    @printf("%s\n", label)
    @printf("  mean A_acc  GPU-fp32 = %.6f   CPU-fp64 = %.6f   Python-fp64 = %.6f\n",
            mean(A_jl_gpu_fp32), mean(A_jl_cpu_fp64), mean(A_py_fp64))
    @printf("  per-trial |GPU-fp32 − CPU-fp64| / |CPU-fp64|:  max = %.4f%%   mean = %.4f%%\n",
            100*maximum(rel_gpu_vs_cpu), 100*mean(rel_gpu_vs_cpu))
    @printf("  per-trial |CPU-fp64 − Python-fp64| / |Python-fp64|:  max = %.4f%%   mean = %.4f%%\n",
            100*maximum(rel_cpu_vs_py), 100*mean(rel_cpu_vs_py))
end


# ── §2.7 item 3: substep ordering review ─────────────────────────────────
# gpu_control.jl lines 87-98: every substep reads (B, F, A) at the top, then
# writes (B, F, A) at the bottom — i.e. simultaneous Euler update, identical
# to JAX's `y_inner + sub_dt * drift_jax(y_inner, …)`. If µ(B,F) read stale
# values, the cumulative drift over 5376 substeps would dwarf the 3e-6
# relative agreement seen above, so the bit-for-bit test is also empirical
# evidence that the substep ordering is correct.
println("\n" * "="^78)
println("§2.7 item 3:  substep-ordering — kernel reads CURRENT (B,F,A) each substep")
println("="^78)
println("  gpu_control.jl:87-98 layout:")
println("    READ:  F_dev, mu_bif, a_factor_B, a_factor_F  (use current B,F,A)")
println("    READ:  drift_B, drift_F, drift_A              (use current B,F,A)")
println("    WRITE: B = B + sub_dt*drift_B                 (last 3 lines)")
println("           F = F + sub_dt*drift_F")
println("           A = A + sub_dt*drift_A")
println("  → simultaneous update, no stale-read hazard. Matches JAX semantics.")
println("  Empirical confirmation: 3-way fp32/fp64 agreement above is impossible")
println("  with stale reads compounding over 5376 substeps.")

File: DEPRECIATED_NON_FUNCTIONAL_CODE/version_2_Julia_DEPRECIATED_NOT WORKING/tools/test_cost_at_theta0.py
#!/usr/bin/env python
"""Bit-for-bit cost comparison Python vs Julia — Python side.

Standing-hypothesis test from julia_fsa_writeup.pdf §2.7 item 1:
fix θ (start with θ = 0), use a saved CRN noise grid, print the
per-trial A_acc / B_acc / barrier_acc that the Python cost_fn
computes. The Julia twin script then reads the same noise and prints
the Julia kernel's values; the comparison localises any A-cost
divergence.

Run from version_2/:
    cd version_2 && PYTHONPATH=.:.. python ../version_2_Julia/tools/test_cost_at_theta0.py

Outputs:
    /tmp/crn_noise_seed42.npy     — (N_INNER, N_STEPS, 3) fp64 Wiener increments
    /tmp/python_cost_at_theta0.npz — per-trial A/B/F/Phi/barrier accumulators
"""

from __future__ import annotations

import os
os.environ.setdefault('JAX_ENABLE_X64', 'True')

import numpy as np
import jax
import jax.numpy as jnp

from models.fsa_high_res._dynamics import (
    TRUTH_PARAMS, drift_jax, diffusion_state_dep,
)
from models.fsa_high_res.control import _make_schedule, INIT_STATE


# ── Config — must match the Julia twin ──────────────────────────────────
OUT_DIR      = "/tmp"
N_INNER      = 32
T_DAYS       = 14
BINS_PER_DAY = 96
N_STEPS      = T_DAYS * BINS_PER_DAY            # 1344
DT           = 1.0 / BINS_PER_DAY               # 15-min outer step
N_SUBSTEPS   = 4
N_ANCHORS    = 8
F_MAX        = 0.40
PHI_MAX      = 3.0
PHI_DEFAULT  = 1.0
SEED         = 42


# ── Step 1: generate fp64 noise and save ────────────────────────────────
rng    = np.random.default_rng(SEED)
wiener = rng.standard_normal((N_INNER, N_STEPS, 3))   # fp64
noise_path = f"{OUT_DIR}/crn_noise_seed{SEED}.npy"
np.save(noise_path, wiener)
print(f"Saved noise → {noise_path}  shape={wiener.shape}  dtype={wiener.dtype}")


# ── Step 2: build em_step + per-trial diagnostic cost in fp64 ───────────
rbf, schedule_from_theta = _make_schedule(
    n_steps=N_STEPS, dt=DT, n_anchors=N_ANCHORS,
    Phi_default=PHI_DEFAULT, Phi_max=PHI_MAX,
)

p_jax  = {k: jnp.asarray(float(v)) for k, v in TRUTH_PARAMS.items()}
sub_dt = DT / float(N_SUBSTEPS)
sqrt_dt = jnp.sqrt(DT)


@jax.jit
def em_step(y, Phi_t, noise_3d):
    def sub_body(y_inner, _):
        return y_inner + sub_dt * drift_jax(y_inner, p_jax, Phi_t), None
    y_det, _ = jax.lax.scan(sub_body, y, jnp.arange(N_SUBSTEPS))
    sigma_y  = diffusion_state_dep(y_det, p_jax)
    y_pred   = y_det + sigma_y * sqrt_dt * noise_3d
    B_pred, F_pred, A_pred = y_pred[0], y_pred[1], y_pred[2]
    B_next = jnp.where(B_pred < 0.0, -B_pred,
                        jnp.where(B_pred > 1.0, 2.0 - B_pred, B_pred))
    F_next = jnp.abs(F_pred)
    A_next = jnp.abs(A_pred)
    return jnp.array([B_next, F_next, A_next])


init_arr = jnp.array([INIT_STATE['B'], INIT_STATE['F'], INIT_STATE['A']])


@jax.jit
def per_trial_components(theta, w_seq):
    """Return (A_acc, B_acc, F_acc, Phi_acc, barrier_acc) for one MC trial.

    Same accumulator semantics as the production cost_fn in
    version_2/models/fsa_high_res/control.py (PRE-step state, not POST-step).
    """
    Phi_arr = schedule_from_theta(theta)

    def step(carry, k):
        y, A_acc, B_acc, F_acc, Phi_acc, barrier_acc = carry
        Phi_t  = Phi_arr[k]
        y_next = em_step(y, Phi_t, w_seq[k])
        # PRE-step accumulation (matches Python's production cost_fn).
        B_acc       = B_acc + y[0] * DT
        F_acc       = F_acc + y[1] * DT
        A_acc       = A_acc + y[2] * DT
        Phi_acc     = Phi_acc + Phi_t * Phi_t * DT
        barrier_acc = barrier_acc + jnp.maximum(y[1] - F_MAX, 0.0) ** 2 * DT
        return (y_next, A_acc, B_acc, F_acc, Phi_acc, barrier_acc), None

    init_carry = (init_arr,
                   jnp.float64(0.0), jnp.float64(0.0),
                   jnp.float64(0.0), jnp.float64(0.0), jnp.float64(0.0))
    (_, A_acc, B_acc, F_acc, Phi_acc, barrier_acc), _ = jax.lax.scan(
        step, init_carry, jnp.arange(N_STEPS),
    )
    return A_acc, B_acc, F_acc, Phi_acc, barrier_acc


per_trial_batched = jax.vmap(per_trial_components, in_axes=(None, 0))


# ── Step 3: evaluate at two θ values ────────────────────────────────────
# θ=0   → Φ ≡ Phi_default = 1.0 (canonical Banister baseline)
# θ_RO  → recovery → overload (anti-symmetric, writeup §2.4 bimodal-test
#         spec: low early, high late). At Phi_max=3, c_Phi=logit(1/3)≈
#         -0.693, so per-anchor sigmoid raw values:
#           θ=-1.5 → raw≈-2.19 → Φ ≈ 0.30
#           θ= 0.0 → raw≈-0.69 → Φ ≈ 1.00
#           θ=+1.5 → raw≈+0.81 → Φ ≈ 2.07
#         Pattern: rest then overload halves of the horizon.
theta0  = jnp.zeros(N_ANCHORS, dtype=jnp.float64)
theta_RO = jnp.array([-1.5, -1.5, -1.5, 0.0, 0.0, +1.5, +1.5, +1.5],
                      dtype=jnp.float64)

fixed_w = jnp.asarray(wiener, dtype=jnp.float64)

results = {}
for label, theta in [("theta0", theta0), ("theta_RO", theta_RO)]:
    A_accs, B_accs, F_accs, Phi_accs, barrier_accs = per_trial_batched(theta, fixed_w)
    Phi_arr = np.asarray(schedule_from_theta(theta))
    results[label] = dict(
        A_acc       = np.asarray(A_accs),
        B_acc       = np.asarray(B_accs),
        F_acc       = np.asarray(F_accs),
        Phi_acc     = np.asarray(Phi_accs),
        barrier_acc = np.asarray(barrier_accs),
        Phi_arr     = Phi_arr,
        theta       = np.asarray(theta),
    )

    print()
    print("=" * 72)
    print(f"Python cost at {label}  (fp64, {N_INNER} CRN trials, T={T_DAYS}d, h=15min)")
    print("=" * 72)
    print(f"  θ           = {np.asarray(theta)}")
    print(f"  Φ(t) summary: min={Phi_arr.min():.3f}  mean={Phi_arr.mean():.3f}  "
          f"max={Phi_arr.max():.3f}")
    Phi_anchors = Phi_arr[::N_STEPS // (N_ANCHORS - 1)][:N_ANCHORS]
    print(f"  Φ at ~anchor times: {Phi_anchors}")
    print()
    print(f"  per-trial A_acc:        mean = {results[label]['A_acc'].mean():.6f}   "
          f"std = {results[label]['A_acc'].std():.6f}")
    print(f"  per-trial F_acc:        mean = {results[label]['F_acc'].mean():.6f}   "
          f"std = {results[label]['F_acc'].std():.6f}")
    print(f"  per-trial barrier_acc:  mean = {results[label]['barrier_acc'].mean():.6f}   "
          f"std = {results[label]['barrier_acc'].std():.6f}")
    print()
    print(f"  cost (-A + barrier):    mean = "
          f"{(-results[label]['A_acc'] + results[label]['barrier_acc']).mean():.6f}")
    print(f"  Mean ∫A/T:              "
          f"{results[label]['A_acc'].mean() / (N_STEPS * DT):.4f}")

# ── Step 4: save per-trial accumulators for the Julia twin ──────────────
out_path = f"{OUT_DIR}/python_cost_at_theta0.npz"
np.savez(out_path,
         # θ=0 results (legacy keys preserved for the existing Julia driver)
         A_acc       = results["theta0"]["A_acc"],
         B_acc       = results["theta0"]["B_acc"],
         F_acc       = results["theta0"]["F_acc"],
         Phi_acc     = results["theta0"]["Phi_acc"],
         barrier_acc = results["theta0"]["barrier_acc"],
         # θ_RO (recovery→overload) results
         A_acc_RO       = results["theta_RO"]["A_acc"],
         B_acc_RO       = results["theta_RO"]["B_acc"],
         F_acc_RO       = results["theta_RO"]["F_acc"],
         Phi_acc_RO     = results["theta_RO"]["Phi_acc"],
         barrier_acc_RO = results["theta_RO"]["barrier_acc"],
         theta_RO       = results["theta_RO"]["theta"],
         Phi_arr_RO     = results["theta_RO"]["Phi_arr"],
)
print(f"\nSaved per-trial results → {out_path}")
